x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -946,22 +988,24 @@ private PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan, PhysicalPl
     public record PreAnalysisResult(
         Set<String> fieldNames,
         Set<String> wildcardJoinIndices,
-        IndexResolution indices,
+        Map<IndexPattern, IndexResolution> indexResolution,
         Map<String, IndexResolution> lookupIndices,
         EnrichResolution enrichResolution,
         InferenceResolution inferenceResolution,
         TransportVersion minimumTransportVersion
     ) {
 
         public PreAnalysisResult(Set<String> fieldNames, Set<String> wildcardJoinIndices) {
-            this(fieldNames, wildcardJoinIndices, null, new HashMap<>(), null, InferenceResolution.EMPTY, null);
+            this(fieldNames, wildcardJoinIndices, Map.of(), new HashMap<>(), null, InferenceResolution.EMPTY, TransportVersion.current());
         }
 
-        PreAnalysisResult withIndices(IndexResolution indices) {
+        PreAnalysisResult withIndices(IndexPattern indexPattern, IndexResolution indices) {
+            Map<IndexPattern, IndexResolution> newIndexResolution = new HashMap<>(this.indexResolution);
```

Comment:
NIT: this structure is mutable already (see lookupIndices). I would not copy the hash map here and just modify existing one. 
Related to https://github.com/elastic/elasticsearch/pull/136686

Reply:
OK. I changed it to follow the pattern used by LookupIndices, and leave the rest of the changes to #136686.